eth/handler.go

```diff
@@ -288,22 +292,74 @@ func newHandler(config *handlerConfig) (*handler, error) {
 	return h, nil
 }
 
+// protoTracker tracks the number of active protocol handlers.
+func (h *handler) protoTracker() {
+	defer h.wg.Done()
+	var (
+		wg       sync.WaitGroup
+		done     = make(chan struct{})
+		active   int
+		quitting bool
+	)
+	// Start goroutine to track active protocol handlers.
+	wg.Add(1)
+	go func() {
+		for {
+			select {
+			case <-h.handlerStartCh:
+				active++
+			case <-h.handlerDoneCh:
+				active--
+			case <-done:
+				quitting = true
+			}
+			// Only check if active is 0 after quitSync has been
+			// closed. This ensures that in case active drops to 0
+			// naturally, this tracker routine stays operational.
+			if quitting && active == 0 {
+				wg.Done()
+				return
+			}
+		}
+	}()
+	<-h.quitSync
+	done <- struct{}{}
+
+	// Wait until all handlers finish.
+	wg.Wait()
+}
+
+// incHandlers signals to increment the number of active handlers if not
+// quitting.
+func (h *handler) incHandlers() bool {
+	select {
+	case h.handlerStartCh <- struct{}{}:
+		return true
+	case <-h.quitSync:
+		return false
+	}
```

Comment:
See my comment above. I am not sure that it's guaranteed that it uses the quitSync branch just because it's available. We want to make it so that as soon as quitSync is closed, it must only use the `return false` branch

Reply:
I was following this pattern: https://github.com/ethereum/go-ethereum/blob/c2db667c8f824ec0d025cbde95e6aaad67fc10ba/eth/sync.go#L92-L99

If we want to ensure false is _always_ returned, would something like this work?

```suggestion
        select {
        case <-h.quitSync:
                return false
        default:
        }
	select {
	case h.handlerStartCh <- struct{}{}:
		return true
	case <-h.quitSync:
		return false
	}
```